core/trino-main/src/main/java/io/trino/connector/WorkerDynamicCatalogManager.java

```diff
@@ -91,29 +106,63 @@ public void ensureCatalogsLoaded(Session session, List<CatalogProperties> expect
             return;
         }
 
-        catalogsUpdateLock.lock();
+        catalogLoadingLock.lock();
         try {
             if (stopped) {
                 return;
             }
 
-            for (CatalogProperties catalog : getMissingCatalogs(expectedCatalogs)) {
-                checkArgument(!catalog.getCatalogHandle().equals(GlobalSystemConnector.CATALOG_HANDLE), "Global system catalog not registered");
-                CatalogConnector newCatalog = catalogFactory.createCatalog(catalog);
-                catalogs.put(catalog.getCatalogHandle(), newCatalog);
-                log.debug("Added catalog: " + catalog.getCatalogHandle());
+            List<CatalogProperties> missingCatalogs = getMissingCatalogs(expectedCatalogs);
+            missingCatalogs.forEach(catalog -> checkArgument(!catalog.getCatalogHandle().equals(GlobalSystemConnector.CATALOG_HANDLE), "Global system catalog not registered"));
+            List<ListenableFuture<Void>> loadedCatalogs = Futures.inCompletionOrder(
+                    missingCatalogs.stream()
+                            .map(catalog ->
+                                    Futures.submit(() -> {
+                                        catalogs.computeIfAbsent(catalog.getCatalogHandle(), ignore -> {
+                                            CatalogConnector newCatalog = catalogFactory.createCatalog(catalog);
+                                            log.debug("Added catalog: " + catalog.getCatalogHandle());
+                                            return newCatalog;
+                                        });
+                                    }, executor))
+                            .collect(toImmutableList()));
+
+            Deque<Throwable> catalogLoadingExceptions = new LinkedList<>();
+            for (ListenableFuture<Void> loadedCatalog : loadedCatalogs) {
+                try {
+                    // Timeout much longer than is expected to load a catalog
+                    // arbitrarily 3x TaskUpdateRequest timeout
+                    loadedCatalog.get(30, TimeUnit.SECONDS);
+                }
+                catch (ExecutionException e) {
+                    if (e.getCause() != null) {
+                        catalogLoadingExceptions.add(e.getCause());
+                    }
+                    else {
+                        catalogLoadingExceptions.add(e);
+                    }
+                }
+                catch (TimeoutException | InterruptedException e) {
```

Comment:
The thread interrupted state is being swallowed here. It needs to be restored or the method needs to throw an exception.

Reply:
Not sure I know enough about restoring the interrupted thread here to know how I might do that. Right now the behavior is to put the exception in the list and throw if there are any exceptions loading any of the catalogs. Is that sub optimal? Or should I just throw immediately in this case (and with what runtime exception)?

Reply:
Believe this is handled now.